cloud-controller-manager/do/firewall_controller_cache.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2020 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"sync"
+
+	"github.com/digitalocean/godo"
+	"k8s.io/klog/v2"
+)
+
+// firewallCache stores a cached firewall and mutex to handle concurrent access.
+type firewallCache struct {
+	sync.RWMutex
+	firewall *godo.Firewall
+	isSet    bool
```

Comment:
Unfortunately not because the firewall we periodically try to fetch from the API could be absent, and we still need to be able to signal to cache users that the cache has been refreshed. If we relied on `firewall` being nil only, we could run into a scenario where reach out to the API on every Service reconcile, leading to a fair number of extra requests. For instance, if we keep failing to create the firewall continuously, the cache would always be considered empty (because `firewall` never becomes non-nil), so we'd try to fetch it again and again until the firewall can eventually be created.

Admittedly it's a bit of an edge case (though I ran into it during tests when the implementation had a subtle bug). I think it's still good to minimize the number of requests we do.

I'll push a small update to comment the field.

Reply:
Gotcha, didn't see that case.